rundeckapp/grails-app/services/rundeck/services/ProjectService.groovy

```diff
@@ -1721,13 +1717,61 @@ class ProjectService implements InitializingBean, ExecutionFileProducer, EventPu
      */
     @GrailsCompileStatic
     DeleteResponse deleteProject(IRundeckProject project, IFramework framework, AuthContext authContext, String username) {
-        def result = new DeleteResponse(success: false)
+        log.info("Requested deletion of project ${project.name}")
+
+        if(framework.getFrameworkProjectMgr().isFrameworkProjectDisabled(project.name)) {
+            return new DeleteResponse(
+            success: false,
+            error: "Cannot delete an already disabled project.")
+        }
+
+        if(!configurationService.getBoolean("projectService.deferredProjectDelete", false)) {
+            return deleteProjectInternal(project, framework, authContext, username)
+        }
+
+        try {
+            Project.withNewTransaction {
+                framework.getFrameworkProjectMgr().disableFrameworkProject(project.name)
+            }
+        } catch (UnsupportedOperationException e) {
+            log.warn("Could not disable project. Aborting project delete operation: ${e.getMessage()}", e)
+            throw new UnsupportedOperationException("Could not delete project. Project disabling not supported by framework: ${e.getMessage()}", e)
+        }
+
+        log.info("Deferring deletion of project ${project.name} to background task.")
+
+        def promise = Promises.task {
+            return Project.withNewTransaction {
```

Comment:
Move `Project.withNewTransaction` to Gorm data provider implementation

Reply:
This process must happen in a separate transaction so it works correctly across multiple threads, and spans many data domains so i cannot move it inside the data provider. I can remove the domain class reference though